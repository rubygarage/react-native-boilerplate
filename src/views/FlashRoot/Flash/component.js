import React from 'react';
import PropTypes from 'prop-types';

import Alert from 'views/shared/Alert';

import useFlashMethods from './hooks/useFlashMethods';

const Flash = (props) => {
  const { flashProps } = props;

  const { onHideFlash, onActionClick } = useFlashMethods(props);

  return (
    <Alert
      {...flashProps}
      onCloseClick={flashProps.isCloseable ? onHideFlash : undefined}
      onActionClick={onActionClick}
    />
  );
};

Flash.propTypes = {
  doAnimation: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired,
  flashProps: PropTypes.shape({
    action: PropTypes.shape({}),
    isCloseable: PropTypes.bool,
  }),
  lifetime: PropTypes.number,
};

Flash.defaultProps = {
  lifetime: undefined,
  flashProps: {},
};

export default Flash;
