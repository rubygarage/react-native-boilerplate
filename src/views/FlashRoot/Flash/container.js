import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { hideFlash as hideFlashAction } from 'state/flash/actions';
import Alert from 'views/shared/Alert';

class Flash extends Component {
  componentDidMount() {
    const { lifetime } = this.props;
    if (lifetime) {
      this.timerHandle = setInterval(this.onHideFlash, lifetime);
    }
  }

  componentWillUnmount() {
    clearInterval(this.timerHandle);
  }

  onHideFlash = () => {
    const { hideFlash, id, doAnimation } = this.props;
    hideFlash(id);
    doAnimation();
  }

  onActionClick = () => {
    const { dispatch, hideFlash, id, flashProps: { action } } = this.props;
    hideFlash(id);
    dispatch(action);
  }

  render() {
    const { flashProps } = this.props;
    const { isCloseable } = flashProps;
    return (
      <Alert
        {...flashProps}
        onCloseClick={isCloseable ? this.onHideFlash : undefined}
        onActionClick={this.onActionClick}
      />
    );
  }
}

Flash.propTypes = {
  hideFlash: PropTypes.func.isRequired,
  doAnimation: PropTypes.func.isRequired,
  dispatch: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired,
  flashProps: PropTypes.shape({
    action: PropTypes.shape({}),
    isCloseable: PropTypes.bool,
  }).isRequired,
  lifetime: PropTypes.number,
};

Flash.defaultProps = {
  lifetime: undefined,
};

const mapDispatchToProps = (dispatch) => ({
  ...bindActionCreators({ hideFlash: hideFlashAction }, dispatch),
  dispatch,
});

export { Flash as FlashContainer };
export default connect(null, mapDispatchToProps)(Flash);
