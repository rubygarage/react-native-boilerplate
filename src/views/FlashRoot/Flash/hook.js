import { useRef, useEffect } from 'react';
import { useDispatch } from 'react-redux';

import { hideFlash } from 'state/flash/actions';

const useContainer = (props) => {
  const timerHandle = useRef();

  const dispatch = useDispatch();

  const {
    id,
    flashProps: { action },
    doAnimation,
    lifetime,
  } = props;

  const onHideFlash = () => {
    dispatch(hideFlash(id));
    doAnimation();
  };

  const onActionClick = () => {
    dispatch(hideFlash(id));
    dispatch(action);
  };

  useEffect(() => {
    if (lifetime) {
      timerHandle.current = setInterval(onHideFlash, lifetime);
    }

    return () => {
      clearInterval(timerHandle.current);
    };
  }, []);

  return {
    onHideFlash,
    onActionClick,
  };
};

export default useContainer;
