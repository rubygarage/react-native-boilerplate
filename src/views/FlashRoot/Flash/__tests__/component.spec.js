import React from 'react';
import { shallow } from 'enzyme';

import Flash from '../component';

jest.useFakeTimers();

jest.mock('../hook', () => ({
  __esModule: true,
  default: () => ({
    onHideFlash: jest.fn(),
    onActionClick: jest.fn(),
  }),
}));

describe('Flash component', () => {
  const props = {
    doAnimation: jest.fn(),
    id: 'test_id',
    flashProps: {
      action: { type: 'TEST' },
      isCloseable: true,
      title: 'Some',
      type: 'info',
    },
    lifetime: null,
  };

  const component = shallow(<Flash {...props} />);

  it('renders correctly', () => {
    expect(component).toMatchSnapshot();
  });

  it('renders correctly when isCloseable equals false', () => {
    component.setProps({
      ...props,
      flashProps: {
        ...props.flashProps,
        isCloseable: false,
      },
    });

    expect(component).toMatchSnapshot();
  });
});
