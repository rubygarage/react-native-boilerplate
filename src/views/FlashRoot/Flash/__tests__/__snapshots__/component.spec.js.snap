// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flash component renders correctly 1`] = `
<Alert
  action={
    Object {
      "type": "TEST",
    }
  }
  isCloseable={true}
  onActionClick={[MockFunction]}
  onCloseClick={[MockFunction]}
  title="Some"
  type="info"
/>
`;

exports[`Flash component renders correctly when isCloseable equals false 1`] = `
<Alert
  action={
    Object {
      "type": "TEST",
    }
  }
  isCloseable={false}
  onActionClick={[MockFunction]}
  title="Some"
  type="info"
/>
`;
